📖 Proposal external CA and cert sign

moh2a opened the pull request:
**What this PR does / why we need it**:
This PR add a proposal fo fixe this [issue](https://github.com/kubernetes-sigs/cluster-api/issues/12585)

/area documentation

moh2a committed:
docs: proposal for external ca and cert sign

moh2a commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,150 @@
+## Upgrade Strategy
+
+No special upgrade plan is required for this feature. Enabling external CA and external certificate signing does not break existing Cluster API behavior. Clusters that use the default internal CA workflow will remain fully functional.
+
```
I didn't think about switching from internal CA to external CA. If you want to make this change, you'll need to ask yourself whether you want to keep the same CAs or use new ones. If you want to use new ones (I think this is the case for you, as your current CAs may already have been exposed), you will need to perform a rotation, which is not trivial: https://kubernetes.io/docs/tasks/tls/manual-rotation-of-ca-certificates/

neolit123 commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,150 @@
+
+## Upgrade Strategy
+
+No special upgrade plan is required for this feature. Enabling external CA and external certificate signing does not break existing Cluster API behavior. Clusters that use the default internal CA workflow will remain fully functional.
```
here you'd have to cover how a cluster can enabled the feature on upgrade and roll it back on failure.

moh2a replied:
Since CAPI does not manage CA rotation automatically, the externalCA feature must be treated as immutable. Enabling it on an existing cluster is not supported. In the future, if CAPI adds support for automated CA rotation, it may be possible to make this field mutable, allowing safe transitions between internal and external CA workflows.

neolit123 reviewed (commented):
i think this is something that has been desired,, which is clear from user stories, and it can ease CA rotation. but the implementation details need a lot more work and discussion here. i think also required changes in the default controllers should be pointed out.

as it stands, this seems to me more as an idea doc and less like a full CAEP.

moh2a committed:
Update docs/proposals/images/external-cert-sign/workflow.plantuml

Co-authored-by: Christian Schlotter <christi.schlotter@gmail.com>

moh2a commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,150 @@
+
+### Certificates necessary to Cluster API
+
+Certificates such as the etcd client certificate and the admin certificate for the kubeconfig are normally generated by CAPI. In external CA mode, these certificates are still needed, but their provisioning and lifecycle management become the responsibility of the user. Users can leverage their own certificate management system, such as cert-manager or any other certificate management solution, t ...
```
The client certificate keys remain stored in Kubernetes Secrets even in external CA mode. This proposal only allows CAPI to operate without storing, using or validating the CA keys in the management cluster, cloud init or in a workload cluster.

moh2a replied:
The user must provide the client certificates directly on the nodes using their preferred mechanism.

moh2a committed:
Update docs/proposals/20250908-external-certificate-signing.md

Co-authored-by: Lubomir I. Ivanov <neolit123@gmail.com>

moh2a commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,150 @@
+### Kubeadm Bootstrap: ClusterConfiguration and InitConfiguration
+
+When using Kubeadm bootstrap, the command kubeadm certs generate-csr requires a configuration file containing ClusterConfiguration and/or InitConfiguration. Currently these files are only provided to the initialization control plane.
+These configurations must be provided to joining control plane nodes because they contain the necessary information to generate Certificate Signing Requests (CSRs) for their components. The joining control plane nodes will generate CSRs locally and submit them to the external CA or user-provided signing system.
```
The init/ClusterConfiguration for joining control plane nodes would be transferred using the same mechanism as for the initial control plane node (using cloud init files)

The responsibilities of CAPI end at enabling external CA mode and signaling that CA keys are managed externally. Generation of CSRs from the nodes and their submission to an external signer are fully the responsibility of the user.

Users can leverage their own mechanisms, such as pre-kubeadm scripts, cloud-init, or any other automation, to generate the CSRs, submit them to the external CA, and retrieve the signed certificates. 

This approach ensures that CAPI remains agnostic to the external signing process, while giving users full flexibility to implement their desired certificate management workflow.

neolit123 replied:
i see, so this CAEP proposes to disable the internal use of CA keys in CAPI by enabling `externalCA`, but the CSR generation, signing of certs and keys and writing them in secrets will be completely external. above, i also ask who is going to write them from secrets to node disks.

moh2a committed:
doc: improve proposal with feature flag, certificate format and upgrade strategy

moh2a committed:
doc: fill history

lentzi90 commented on docs/proposals/20250908-external-certificate-signing.md:
````diff
@@ -0,0 +1,167 @@
+data:
+  ca.crt: <base64-encoded-CA-certificate>
+```
+  CAPI should wait for the CA certificates if they are not already provided. CAPI must not adopt these Secrets, as they are the responsibility of the user.”
````
```suggestion
  CAPI should wait for the CA certificates if they are not already provided. CAPI must not adopt these Secrets, as they are the responsibility of the user.
```

moh2a committed:
Update docs/proposals/20250908-external-certificate-signing.md

Co-authored-by: Lennart Jern <lennart.jern@gmail.com>

moh2a committed:
Update docs/proposals/20250908-external-certificate-signing.md

Co-authored-by: Lennart Jern <lennart.jern@gmail.com>

moh2a committed:
docs: detail conditions and upgrade strategy

moh2a commented:
Hi @neolit123 , @sbueringer 
I hope you’ve been doing well. Sorry for my delay I’ve been quite busy over the past few months. I’ve answered all of your questions in the proposal but please don’t hesitate to ask if you need any further clarification

Since you’re focused on other more pressing matters perhaps we could postpone this proposal for now as most of the points raised aren’t blocking. The exception is this PR: https://github.com/kubernetes-sigs/cluster-api/pull/12603. As it’s more of a fix than a new feature it might not strictly need a full proposal. Maybe we could focus on that one for now as it seems relatively straightforward. What do you think?

sbueringer commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,167 @@
+
+### External CA
+
+An External Certificate Authority (CA) is a CA that exists outside the Cluster API (CAPI) management plane. Instead of having Cluster API generate and manage its own self-signed CA for cluster components (such as kube-apiserver, kubelet, and etcd), an external CA can be integrated to provide trusted root certificates. This is typically used in environments where certificate lifecycles are central ...
```
The ca is not providing trusted root certs it has to provide all certs that we generate from these cas right?

sbueringer commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,167 @@
+
+The etcd client certificate should be provided by the user in a secret named <cluster-name>-apiserver-etcd-client
+The kubeconfig admin certificate should be provided by the user in a secret named <cluster-name>-admin
+
```
What about the -kubeconfig secret that almost all our controllers depend on?

moh2a replied:
KCP/CABKP can build it using the secret <cluster-name>-admin

fabriziopandini replied:
Can't we simplify this and let the external tool to manage the kubeconfig secret, including also certificate rotation?

sbueringer commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,167 @@
+
+### Kubeadm Bootstrap: ClusterConfiguration and InitConfiguration
+
+When using Kubeadm bootstrap, the command kubeadm certs generate-csr requires a configuration file containing ClusterConfiguration and/or InitConfiguration. Currently these files are only provided to the initialization control plane.
```
ToDo: check current state

sbueringer commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,167 @@
+
+### Non-Goals/Future Work
+
+- Cluster API will not provide a built-in external CA service; users remain responsible for setting up and managing the external CA.
```
TODO: non-goal get rid of kubeconfig secret + etcd client cert/key secret

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,167 @@
+
+### External certificate signing
+
+External certificate signing is the process of delegating certificate issuance to a signing service instead of relying on self-signed certificates generated by Cluster API or a [bootstrap provider](https://cluster-api.sigs.k8s.io/reference/glossary.html#bootstrap-provider). With this approach, the Cluster API user becomes responsible for providing the signed certificates, ensuring their signing b ...
```
I think we should not talk about " the Cluster API user becomes responsible for providing the signed certificates", because in many plaform/K8s distribution built on top of Cluster API automation is paramount (there should not be human tasks, especially when those iterations might require involvement of different teams/people with different responsibilities)

What about instead using the following wording: "Cluster API delegates to external systems (or users) the responsibility for providing the signed certificates"

(same comments applies to other parts in the doc, I think we should align all of them to the idea that a system or a human can take care of tasks not performed by humans)

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,167 @@
+
+## Summary
+
+This proposal introduces the ability for CAPI to integrate with fully external Certificate Authorities (CAs), meaning that CA key never leaves the external system. It also supports external certificate signing, where all certificate issuance is delegated externally. Cluster components generate Certificate Signing Requests (CSRs) that are signed by the external CA, providing a centralized and cons ...
```
Am I correct is say that this proposal aims to enable external certificate signing for external CA?

This will make a direct link to the glossary above, and drop the need to introduce the concept of "fully" external CA. In other words, at the end CAPI will support

- External CA without external certificate signing  (already supported today)
- External CA with external certificate signing  (this proposal)

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,167 @@
+### Goals
+
+- Enable Cluster API to integrate with fully external Certificate Authorities (CAs) where CA private keys never leave the external system.
+- Support external certificate signing for cluster components (control plane and nodes), delegating certificate issuance to an external CA or signing service.
```
I would phrase this "Make external certificate signing for cluster components possible", because external components/workflow are required to fully support this

fabriziopandini reviewed (commented):
Thanks for this PR!
A few general comments.
- I think we should scope down the proposal a little bit, making sure that it is clear for the reader what is in scope and what not (e.g. KCP/kubeadm vs all the bootstrap providers)
- While this proposal defer much to the users, I think we should describe how those manual task could be automated, because Cluster API is often used as a building block for platform (a viable path to automation is paramount)

sbueringer commented on docs/proposals/20250908-external-certificate-signing.md:
````diff
@@ -0,0 +1,167 @@
+```
+// externalCA enables external CA mode for the cluster.
+// When enabled, CA keys are not stored or validated in the management cluster's secrets.
+ExternalCA *bool `json:"externalCA,omitempty"`
````
I think we should not use the feature flag to enable/disable this behavior for all Clusters.
Usually feature gates are eventually locked to true before they are removed.
That would not be an option here if we also use it to enable/disable it for all Clusters

+ If we scope this feature down to KCP/CABPK this should probably move to the KCP object/controller

fabriziopandini replied:
Wandering a little bit if the system can infer this by checking if CA key are provided, similar to what kubeadm does when looking at the etc/kubernetes/pki folder

fabriziopandini replied:
I'm leaning towards avoiding API changes and inferring external CA mode from the missing key

sbueringer commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,167 @@
+The etcd client certificate should be provided by the user in a secret named <cluster-name>-apiserver-etcd-client
+The kubeconfig admin certificate should be provided by the user in a secret named <cluster-name>-admin
+
+### Kubeadm Bootstrap: ClusterConfiguration and InitConfiguration
```
+1 if kubeadm join is able to generate certs with only a JoinConfiguration this should be possible (might require kubeadm changes)

fabriziopandini replied:
The proposal doesn't give any indication on how certification generation on nodes should happen, we need more clarity before taking any decisions.

Please note that sending ClusterConfiguration to the machines is not easy as it seems, because this will require deep and complex investigations on how this might impact how those file should look like and how to handle them when detecting if a rollout is needed or when to perform in-place upgrades (plus the capability to handle file generate before and after this change).
TL;DR; this might impact the size and scope of this proposal in a relevant way

moh2a commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,167 @@
+
+- Enable Cluster API to integrate with fully external Certificate Authorities (CAs) where CA private keys never leave the external system.
+- Support external certificate signing for cluster components (control plane and nodes), delegating certificate issuance to an external CA or signing service.
+- Allow users to implement their own PKI workflow, including certificate provisioning, signing, and rotation, while maintaining secure cluster bootstrap and lifecycle operations.
```
Kubelet Certificates: For worker nodes, the Kubelet natively handles its own certificate lifecycle by generating a Certificate Signing Request (CSR) when certificates are nearing expiration. This makes an external custom signer the appropriate mechanism for this layer (https://github.com/kubernetes-sigs/cluster-api/pull/12739#discussion_r2792559626).

Control Plane Certificates: For all control plane components, KubeadmControlPlane (KCP) already supports automated rotation through machine rollouts. This ensures that certificates are refreshed during the standard node lifecycle operations, as documented in the [CAPI certificate rotation guide](https://cluster-api.sigs.k8s.io/tasks/certs/auto-rotate-certificates-in-kcp#automatically-rotating-certificates-using-kubeadm-control-plane-provider).

moh2a commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,167 @@
+- Key transit over cloud-init: The CA private keys travel over cloud-init user data during control plane provisioning, which could be intercepted or leaked.
+- Privilege escalation risk: With access to CA keys, an attacker can sign admin or superadmin certificates, gaining full control over clusters and workloads.
+
+For reference, native Kubernetes clusters generate CA keys and store them on control plane nodes. This also represents a security risk because compromise of a control plane node could allow signing of highly privileged certificates.
```
By 'native Kubernetes clusters', I refer to standard clusters deployed with kubeadm, where CA keys are generated and stored locally on the control plane nodes by default. I have [updated ](https://github.com/kubernetes-sigs/cluster-api/pull/12739/changes#diff-52fa92389a92fc1c77afaad2be03648839f99ff4e01030afffd789f706986a08R65)the proposal to avoid any confusion.

moh2a committed:
doc: scope to KCP/CABPK, add non-goals + table for impacted certificates

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,239 @@
+- KCP/CABPK users can implement centralized enterprise PKI practices, ensuring strict access control and auditability for certificate issuance.
+- Overall trust in cluster bootstrap and lifecycle operations is increased, mitigating a high-impact attack vector that currently exists in default Kubernetes deployments and, by extension, Cluster API workload cluster.
+
+This proposal addresses a critical security gap, providing users with a mechanism to eliminate CA key exposure while maintaining full Kubernetes functionality.
```
nit
```suggestion
This proposal provides users with a mechanism to eliminate CA key exposure while maintaining full Kubernetes functionality.
```

I will also move this at ~L66 (before enumerating the advantages a user can achieve by eliminating CA exposure)

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,239 @@
+- This proposal does not modify Kubernetes internal certificate signing behavior beyond enabling delegation to an external authority.
+- The automated rotation of the Certificate Authorities themselves is not in scope for this proposal.
+- This proposal does not protect against all damages if an attacker gains access to the management cluster, control plane nodes, or cluster admin credentials.
+- This proposal does not protect against all damages if an attacker gains access to the management cluster, control plane nodes, or cluster admin credentials.
```
```suggestion
```
duplicated

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,239 @@
+
+### Goals
+
+- Enable KCP/CABPK to integrate with external Certificate Authorities (CAs) where CA private keys never leave the external system.
```
This goal should be stated in a more generic way, because KCP/CABK are pluggable bootstrap/control plane provider and their behaviour and interactions with CAPI should be defined by the CAPI contral

fabriziopandini replied:
Same comment apply in a few other places in this doc.
We should solve the problem in a generic way, and move all the references to KCP, CABPK (and kubeadm) int a single paragraph where we explain how the generic approach apply to those components

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,239 @@
+
+#### Cluster API Contract and Kubeconfig
+
+The admin certificate used for the cluster kubeconfig is a core Cluster API requirement. This proposal adheres to the [Cluster API Control Plane Contract](https://cluster-api.sigs.k8s.io/developer/providers/contracts/control-plane#cluster-kubeconfig-management) by delegating the issuance of the `<cluster-name>-admin` secret to an external system while maintaining standard naming conventions. This ...
```
I'm not entirely sure this proposal adheres to the current contract.

The current contract assigns to the control plane provider the resposnibility to create and maintain (rotate) the Kubeconfig secret, which is not true anymore in external CA mode, where this responsibility is assigned to an external system)

(TL;DR; I think we have to change the contract in https://cluster-api.sigs.k8s.io/developer/providers/contracts/control-plane#cluster-kubeconfig-management for this proposal to explicitly call this out).

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,239 @@
+
+### Certificates necessary to KCP/CABPK
+
+#### Kubeadm and KCP requirements
```
I think that something else must change in CABPK, because it should stop sending keys to the machines as part of the cloud init secret.

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,239 @@
+The management of Certificate Authorities and internal component certificates falls under the scope of KCP and CABPK. KCP will not generate these keys in external CA mode and expects the user to provide the following secrets:
+
+- **Certificate Authorities:** The cluster CA, etcd CA, and front-proxy CA must be provided as secrets.
+- **Etcd Client Certificate:** For clusters using stacked etcd, the user must provide the etcd client certificate in a secret named `<cluster-name>-apiserver-etcd-client`. For external etcd, KCP follows the existing logic for external etcd connectivity.
```
Does the user should take care of the lifecycle for this certificate (e.g. rotation, removal)? if yes, please call this out

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,239 @@
+
+The management of Certificate Authorities and internal component certificates falls under the scope of KCP and CABPK. KCP will not generate these keys in external CA mode and expects the user to provide the following secrets:
+
+- **Certificate Authorities:** The cluster CA, etcd CA, and front-proxy CA must be provided as secrets.
```
Let's also mention that the user should also take care of secret removal

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
````diff
@@ -0,0 +1,239 @@
+data:
+  ca.crt: <base64-encoded-CA-certificate>
+```
+KCP/CABPK should wait for the CA certificates if they are not already provided. KCP/CABPK must not adopt these Secrets, as they are the responsibility of the user.
````
I'm a little bit concerned by 

> KCP/CABPK should wait for the CA certificates 

This implies that we are implementing also an orchestration process between KCP/CABPK and the external CA system, which is something we are not doing e.g in case of custom CA

TBH. I would prefer if this is remove from the proposal, because e.g. we already have other mechanism like the BeforeClusterCreate hooks that provide a similar orchestration workflow.

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,239 @@
+
+The following table summarizes the certificates impacted by the external CA mode:
+
+| Certificate / Key Pair    | Responsibility  | Scope         | Location           | Logic / Mechanism                                |
```
I would drop scope, because IMO it is confusing

e.g. KCP provides certificates in compliance to the [Cluster API contract](https://cluster-api.sigs.k8s.io/developer/providers/contracts/control-plane#cluster-certificate-management), but it seems that only the KubeConfig is managed according to the contract (which is not correct)

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,239 @@
+| **Front Proxy CA**        | External system | KCP           | Management Cluster | Key in external system; Cert in Secret           |
+| **Etcd Client Cert**      | External system | KCP           | Management Cluster | Provided via Secret (for stacked etcd)           |
+| **Kubeconfig Admin Cert** | External system | CAPI Contract | Management Cluster | Provided via Secret                              |
+| **Controlplane Certs**    | External system | CABPK         | Controlplane node  | Provisioned during bootstrap                     |
+| **Kubelet Cert**          | External system | CABPK         | node               | Handled via Kubernetes CSR API and custom signer |
```
I'm not entirely sure about those two lines, because what we are describing here is very specific to a provider/to how kubeadm works and this fact is not surfacing

fabriziopandini replied:
Might be worth to add a generic version of this table, limited to what will be prescribed by the contract), and then a second version that show the specific implementation for KCP/CABPK

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,239 @@
+#### Cluster API Contract and Kubeconfig
+
+The admin certificate used for the cluster kubeconfig is a core Cluster API requirement. This proposal adheres to the [Cluster API Control Plane Contract](https://cluster-api.sigs.k8s.io/developer/providers/contracts/control-plane#cluster-kubeconfig-management) by delegating the issuance of the `<cluster-name>-admin` secret to an external system while maintaining standard naming conventions. This ...
+Certificates such as the etcd client certificate and the admin certificate for the kubeconfig are normally generated by KCP/CABPK. In external CA mode, these certificates are still needed, but their provisioning and lifecycle management become the responsibility of the user.
```
How other CAPI users are expected to do so?

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,239 @@
+The admin certificate used for the cluster kubeconfig is a core Cluster API requirement. This proposal adheres to the [Cluster API Control Plane Contract](https://cluster-api.sigs.k8s.io/developer/providers/contracts/control-plane#cluster-kubeconfig-management) by delegating the issuance of the `<cluster-name>-admin` secret to an external system while maintaining standard naming conventions. This ...
+Certificates such as the etcd client certificate and the admin certificate for the kubeconfig are normally generated by KCP/CABPK. In external CA mode, these certificates are still needed, but their provisioning and lifecycle management become the responsibility of the user.
+
+Kubeadm controllers will verify the presence and validity of external certificates using **two distinct conditions** on the `KubeadmControlPlane` resource:
```
I will try to keep it simple and use existing CertificatesAvailable in KCP/CABPK to surface issues

e.g. condition must be true when
- Certificate authorities are present and all have keys (custom CA or self signed)
- Certificates authorities are present and all don't have keys (custom CA or self signed) + user managed client cert are there as well

e.g. condition must be false when
- Failure in generating certificate authorities (self signed)
- Certificates authorities are present but in a mixed mode (some have keys, come other not)
- When in external CA mode, some of the user managed client cert are missing

fabriziopandini commented on docs/proposals/20250908-external-certificate-signing.md:
```diff
@@ -0,0 +1,239 @@
+
+## Upgrade Strategy
+
+The `externalCA` feature can be enabled on an existing cluster that was originally created with the default internal CA, but it requires careful manual steps by the user. Cluster API does not automatically rotate or migrate CAs.
```
I would prefer to rule this out (non goal) to keep the scope of the proposal limited.

fabriziopandini reviewed (commented):
I think we can try to get this proposal moving in the 1.14 timeframe

Main point to be addressed by me:
- The proposal should be phrase in a generic way / in term of contract changes, because KCP/CABPK (and kubeadm) are pluggable components
- I would prefer to remove from the current proposal any sort of orchestration during cluster creation, because this overlaps with e.g the BeforeClusterCreate hook.
- The proposal does not provide any hint on how a CAPI users can implement the process required to make this work (kubeconfig management, certificate provisioning during machine bootstrap) and without this it will be un-usable by most of the CAPI users. We need to provide some guidance + think about a minimal solution for E2E coverage of what is in scope of this proposal

moh2a committed:
docs: Addresses review feedback: proposal should be phrased in terms of contract changes
since KCP/CABPK are pluggable components. certificate provisioning during machine bootstrap.

moh2a commented:
Hi everyone,
I wanted to give an update on this PR. Unfortunately, due to personal and professional commitments lately, I haven't had the time to dedicate to moving this forward. I'm sorry for leaving this hanging for so long.
Additionally, the corresponding proposal hasn't garnered much enthusiasm from the community, so I think it's best to close this PR rather than keep it open indefinitely. That said, if anyone is interested in picking this up, feel free to do so.
I want to sincerely thank the maintainers who took the time to review and provide feedback, it's been appreciated.
Thanks again, and sorry for the inconvenience.
